scripts/include/helpers.inc.php

```diff
@@ -121,45 +121,97 @@ function mergeXliffDomToAtom(DOMDocument $domA, DOMDocument $domB): DOMDocument
 
 function mergeXliffDomToWeblate(DOMDocument $domA, DOMDocument $domB): DOMDocument
 {
-  // Elements in A but not in B -> copy to B
-  $sourcesA = $domA->getElementsByTagName('source');
-  $sourcesB = $domB->getElementsByTagName('source');
+  // Elements in A AND B -> check source string and if different -> replace in B
+  $transUnitsA = $domA->getElementsByTagName('trans-unit');
+  $transUnitsB = $domB->getElementsByTagName('trans-unit');
 
-  foreach ($sourcesA as $source) {
-    $stringsA[] = $source->nodeValue;
+  $i = 0;
+  foreach ($transUnitsA as $transUnit) {
+    if ($transUnitsA->item($i)->hasAttribute('id')) {
+      $transUnitsA->item($i)->setIdAttribute('id', true);
+
+      $idListA[] = $transUnitsA->item($i)->getAttribute('id');
+    }
+
+    $i++;
   }
 
-  foreach ($sourcesB as $source) {
-    $stringsB[] = $source->nodeValue;
+  $i = 0;
+  foreach ($transUnitsB as $transUnit) {
+    if ($transUnitsB->item($i)->hasAttribute('id')) {
+      $transUnitsB->item($i)->setIdAttribute('id', true);
+
+      $idListB[] = $transUnitsB->item($i)->getAttribute('id');
+    }
+
+    $i++;
   }
 
-  $copyToB = array_diff($stringsA, $stringsB);
+  // Items in A that are also in B.
+  $verifyInB = array_intersect($idListA, $idListB);
 
-  foreach($sourcesA as $source) {
-    if (in_array($source->nodeValue, $copyToB)) {
-      // get parent which should be a trans-unit
-      $parent = $source->parentNode;
-      // add trans-unit as child of domB body
-      $bodyB = $domB->getElementsByTagName('body')[0];
-      $bodyB->appendChild($domB->importNode( $parent, true ));
+  // Use trans-unit's id attribute to identify strings. When one is
+  // matched, should also compare the source string target and if
+  // different, replace it in target.
+  $i = 0;
+  foreach ($transUnitsA as $transUnit) {
+    $id = $transUnitsA->item($i)->getAttribute('id');
+
+    if (in_array($id, $verifyInB)) {
+      $node = $domB->getElementById($id);
+
+      // Get 'source' string element from dom B trans-unit.
+      $sourcesB = $node->getElementsByTagName('source');
```

Comment:
Suggestion: this could be turned into function to remove the duplicate code below for `$stringA`.

Reply:
Done.